src/periodic_wave.rs

```diff
@@ -0,0 +1,423 @@
+                        r.len() >= 2,
+                        "RangeError: Real field length should be at least 2"
+                    );
+                    assert!(
+                        // the specs gives this number as a lower bound
+                        // it is implemented here as a upper bound to enable required casting
+                        // without loss of precision
+                        r.len() <= 8192,
+                        "NotSupported: periodic wave of more than 8192 components"
+                    );
+                    let r_len = r.len();
+                    (r, vec![0.; r_len])
+                }
+                (None, Some(i)) => {
+                    assert!(
+                        i.len() >= 2,
+                        "RangeError: Real field length should be at least 2"
+                    );
+                    assert!(
+                        i.len() <= 8192,
+                        // the specs gives this number as a lower bound
+                        // it is implemented here as a upper bound to enable required casting
+                        // without loss of precision
+                        "NotSupported: periodic wave of more than 8192 components"
+                    );
+                    let i_len = i.len();
+                    (vec![0.; i_len], i)
+                }
+                _ => (vec![0., 0.], vec![0., 1.]),
+            };
+
+            (real, imag, !disable_normalization.unwrap_or(false))
+        } else {
+            (vec![0., 0.], vec![0., 0.], true)
+        };
+
+        let wavetable = Self::generate_wavetable(&real, &imag, normalize, TABLE_LENGTH_USIZE);
+
+        Self {
+            wavetable: Arc::new(wavetable),
+        }
+    }
+
+    pub(crate) fn as_slice(&self) -> &[f32] {
+        &self.wavetable[..]
+    }
+
+    // cf. https://webaudio.github.io/web-audio-api/#waveform-generation
+    // note that sines are in the imaginary components
+    //
+    // the current implementation is naive and could be improved using inverse FFT
+    // or table lookup on SINETABLE
+    fn generate_wavetable(reals: &[f32], imags: &[f32], normalize: bool, size: usize) -> Vec<f32> {
+        let mut wavetable = Vec::with_capacity(size);
+        let pi_2 = 2. * PI;
+
+        for i in 0..size {
+            let mut sample = 0.;
+            let phase = pi_2 * i as f32 / size as f32;
+
+            for j in 1..reals.len() {
+                let freq = j as f32;
+                let real = reals[j];
+                let imag = imags[j];
+                let rad = phase * freq;
+                let contrib = real * rad.cos() + imag * rad.sin();
+                sample += contrib;
+            }
+
+            wavetable.push(sample);
+        }
+
+        if normalize {
+            Self::normalize(&mut wavetable);
+        }
+
+        wavetable
+    }
+
+    fn normalize(wavetable: &mut [f32]) {
+        let mut max = 0.;
+
+        for sample in wavetable.iter() {
```

Comment:
Have a look at https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.max
The previous version used a version with `reduce`, but that is also overly complicated

Reply:
Yep, but as far I understand the previous implementation didn't check for absolute values, which is important here because if your signal is `[-0.5, 0.1]` you want your normalization factor to be `2.`, not `10.`. Maybe it could be written in a more functional style though.

Reply:
Yup, like `wavetable.iter().map(|&v| v.abs()).max().unwrap()`